src/dev/lupio/lupio_pic.cc

```diff
@@ -35,6 +35,7 @@
 #include "sim/system.hh"
 
 #define LUPIO_PIC_NSRC      32
+#define LUPIO_PIC_NCPU      32
```

Comment:
Is it required that this and LUPIO_IPI_NCPU are the same? If so, then this is a potential future bug if they become out of sync. If there is some fundamental reason why there is a specific maximum number of CPUs, then that should be specified as a single variable. If it's possible to make this a parameter it would be even better.

Reply:
It's funny because I started writing this comment to explain why the number of cpus was limited to 32 by design, only to realize that it actually wasn't. Unless I've missed something, it looks like LupIO can support any number of CPUs. Each SMP-ready device (timer, pic, ipi) can have its interface duplicated in as many CPUs as there are in the system.

The number of IRQ sources is limited to 32 by design though, because LupIO assumes 32-bit registers across all devices. The 32-bit `pending` register can therefore reflect IRQs from 32 devices max. That explains the `NSRC` value.